libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -699,16 +699,22 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 		}
 		case FunctionType::Kind::SHA3:
 		{
-			TypePointers argumentTypes;
-			for (auto const& arg: arguments)
+			solAssert(arguments.size() == 1, "");
+			solAssert(!function.padArguments(), "");
+			TypePointer const& argType = arguments.front()->annotation().type;
+			solAssert(argType, "");
+			arguments.front()->accept(*this);
+			if (*argType == ArrayType(DataLocation::Memory) || *argType == ArrayType(DataLocation::Memory, true))
```

Comment:
Do we really want to support `keccak256(bytes)` and `keccak256(*) -> keccak256(abi.encodePacked(*)` implicitly?

Reply:
Sorry, I don't really understand. This is an optimization step in case the argument is `bytes memory` (as opposed to `bytes storage` or `bytes calldata`), where we do not have to encode anything. Encoding requires additional memory.

This does not mean that we support anything apart from `keccak256(bytes memory)`.

Reply:
Correction: Encoding requires additional copy operations. The memory is only reserved temporarily.

Reply:
Ok! Can you add a comment to mention this is an optimisation?